### PR TITLE
Improve asserts in MF/FEEval regarding FENothing

### PR DESCRIPTION
**Pull request by peterrum**
FYI @sebproell

### CODE REVIEW

File: include/deal.II/matrix_free/matrix_free.h [hunk 1]
```diff
@@ -1884,6 +1884,24 @@ class MatrixFree : public Subscriptor
...
                       const unsigned int hp_active_fe_index = 0) const;
 
+  /**
+   * Return the category the current batch range of cells was assigned to.
+   * Categories run between the given values in the field
+   * AdditionalData::cell_vectorization_category for non-hp-DoFHandler types
+   * and return the active FE index in the hp-adaptive case.
+   */
+  unsigned int
+  get_cell_range_category(
```
Comment (peterrum) [inline include/deal.II/matrix_free/matrix_free.h:1894]: @Rombur I guess you use `get_cell_category`? Or how do you skip cells with `FENothing`?

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 2]
```diff
@@ -6599,6 +6599,19 @@ FEEvaluation<dim,
   (void)dof_no;
   (void)first_selected_component;
 
+  Assert(
+    this->data->dofs_per_component_on_cell > 0,
+    ExcMessage(
+      "There is nothing useful you can do with an FEEvaluation "
+      "object with FE_Nothing, i.e., without DoFs! If you have passed "
+      "to MatrixFree::reinit() a collection of finite elements also containing "
```
Comment (kronbichler) [inline include/deal.II/matrix_free/fe_evaluation.h:6607]: Can you break this line before 'containing' or, alternatively, use similar widths in all lines of the message?

File: include/deal.II/matrix_free/matrix_free.h [hunk 3]
```diff
@@ -1884,6 +1884,24 @@ class MatrixFree : public Subscriptor
...
+    const std::pair<unsigned int, unsigned int> cell_batch_range) const;
+
+  /**
+   * Return the category on the cells on the two sides of the current batch
```
Comment (kronbichler) [inline include/deal.II/matrix_free/matrix_free.h:1898]:
```suggestion
   * Return the category of the cells on the two sides of the current batch
```

File: include/deal.II/matrix_free/util.h [hunk 4]
```diff
@@ -110,6 +110,13 @@ namespace internal
     inline std::pair<Quadrature<dim - 1>, Quadrature<dim - 1>>
     get_unique_face_quadratures(const Quadrature<dim> &quad)
     {
+      AssertThrow(
+        quad.size() > 0,
+        ExcMessage(
+          "There is nothing useful you can do with a MatrixFree/FEEvaluation "
```
Comment (kronbichler) [inline include/deal.II/matrix_free/util.h:116]: Same here, it would be better to break the line before `MatrixFree`.

**Review by kronbichler — APPROVED** [review]
These assertions are certainly helpful, thank you.

**Comment by peterrum** [comment]
@kronbichler Done!

File: include/deal.II/matrix_free/matrix_free.h (hunk 1 above)
Comment (Rombur) [reply include/deal.II/matrix_free/matrix_free.h:1894]: I use `create_cell_subrange_hp_by_index` to get the cells that have a `fe_index` different than the one I use for `FE_Nothing`

**Review by kronbichler — APPROVED** [review]

**Pull request merged**